File: java/socketd-transport-test/src/test/java/labs/ServerTest.java
package labs;

import org.noear.socketd.SocketD;
import org.noear.socketd.transport.core.EntityMetas;
import org.noear.socketd.transport.core.entity.StringEntity;
import org.noear.socketd.transport.core.listener.EventListener;
import org.noear.socketd.utils.RunUtils;
import org.noear.socketd.utils.StrUtils;

public class ServerTest {
    static final String[] schemas = new String[]{
            "sd:tcp-java",
            "sd:tcp-netty",
            "sd:tcp-smartsocket",//2
            "sd:ws-java",
            "sd:udp-java", //4
            "sd:udp-netty",
            "sd:kcp-java",
    };

    /**
     * 启动服务，给别的客户端调试
     */
    public static void main(String[] args) throws Exception {
        String s1 = schemas[3];
        SocketD.createServer(s1)
                .config(c -> c.port(8602))
                .listen(new EventListener()
                        .doOnOpen(s -> {
                            System.out.println("onOpen: " + s.sessionId());
                        }).doOnMessage((s, m) -> {
                            System.out.println("onMessage: " + m);

                            if (m.isRequest()) {
                                String fileName = m.meta(EntityMetas.META_DATA_DISPOSITION_FILENAME);
                                if (StrUtils.isEmpty(fileName)) {
                                    s.reply(m, new StringEntity("me to!"));
                                } else {
                                    s.reply(m, new StringEntity("file received: " + fileName + ", size: " + m.dataSize()));
                                }
                            }

                            if (m.isSubscribe()) {
                                int size = m.metaAsInt(EntityMetas.META_RANGE_SIZE);
                                for (int i = 1; i <= size; i++) {
                                    s.reply(m, new StringEntity("me to-" + i));
                                }
                                s.replyEnd(m, new StringEntity("welcome to my home!"));
                            }
                        }).doOn("/push", (s, m) -> {
                            if(s.attrHas("push")){
                                return;
                            }

                            s.attrPut("push", "1");

                            while (true) {
                                if (s.attrHas("push") == false) {
                                    break;
                                }

                                s.send("/push", new StringEntity("push test"));
                                RunUtils.runAndTry(() -> Thread.sleep(1000));
                            }
                        }).doOn("/unpush", (s, m) -> {
                            s.attrMap().remove("push");
                        })
                        .doOnClose(s -> {
                            System.out.println("onClose: " + s.sessionId());
                        }).doOnError((s, err) -> {
                            System.out.println("onError: " + s.sessionId());
                            err.printStackTrace();
                        }))
                .start();
    }
}

File: python/socketd/test/labs/ServerTest.py
import asyncio

from socketd import SocketD
from socketd.transport.core.entity.StringEntity import StringEntity
from socketd.utils.LogConfig import log
from socketd.transport.core.listener.EventListener import EventListener


async def main():
    server = await (SocketD.create_server("sd:ws")
                    .listen(EventListener()
                            .do_on_open(lambda s:
                                        # 会话打开时
                                        log.info(s.session_id())
                                        ).do_on_message(lambda s, m:
                                                        # 收到任意消息时（方便做统一的日志打印）
                                                        log.info(m.data_as_string())
                                                        ).do_on("/demo", lambda s, m:
    # 收到"/demo"事件的消息时。如果是请求或订阅？则进行签复
    (m.is_request() or m.is_subscribe()) and s.reply_end(m, StringEntity("And you too."))
                                                                )).start())

    await asyncio.Future()


if __name__ == "__main__":
    asyncio.run(main())